File: tests/test_7_7_2_p_predicates.py
"""
Tests for Task 7.7.2: P Predicate Additions.

Validates startingWith, notStartingWith, endingWith, notEndingWith,
containing, notContaining, and regexp predicates.
See docs/project/changelog/task-7.7.2-p-predicates.adoc.
"""

import pytest
import re
import json

from tinkercat import TinkerCat

# ---------------------------------------------------------------------------
# Mock P predicates
# ---------------------------------------------------------------------------

class P:
    def __init__(self, fn, pattern: str):
        self._fn = fn
        self.pattern = pattern

    def test(self, value) -> bool:
        if value is None:
            return False
        return self._fn(str(value))

    def negate(self) -> "P":
        fn = self._fn
        return P(lambda v: not fn(v), self.pattern)

    @staticmethod
    def starting_with(prefix: str) -> "P":
        return P(lambda v: v.startswith(prefix), prefix)

    @staticmethod
    def not_starting_with(prefix: str) -> "P":
        return P(lambda v: not v.startswith(prefix), prefix)

    @staticmethod
    def ending_with(suffix: str) -> "P":
        return P(lambda v: v.endswith(suffix), suffix)

    @staticmethod
    def not_ending_with(suffix: str) -> "P":
        return P(lambda v: not v.endswith(suffix), suffix)

    @staticmethod
    def containing(sub: str) -> "P":
        return P(lambda v: sub in v, sub)

    @staticmethod
    def not_containing(sub: str) -> "P":
        return P(lambda v: sub not in v, sub)

    @staticmethod
    def regexp(pattern: str) -> "P":
        compiled = re.compile(pattern)
        return P(lambda v: bool(compiled.search(v)), pattern)

# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_starting_with_match():
    assert P.starting_with("Al").test("Alice")

def test_starting_with_no_match():
    assert not P.starting_with("Al").test("Bob")

def test_starting_with_null_safe():
    assert not P.starting_with("Al").test(None)

def test_not_starting_with():
    p = P.not_starting_with("Al")
    assert p.test("Bob")
    assert not p.test("Alice")

def test_ending_with_match():
    assert P.ending_with("son").test("Jackson")

def test_ending_with_no_match():
    assert not P.ending_with("son").test("Alice")

def test_ending_with_null_safe():
    assert not P.ending_with("son").test(None)

def test_not_ending_with():
    p = P.not_ending_with("son")
    assert p.test("Alice")
    assert not p.test("Jackson")

def test_containing_match():
    assert P.containing("li").test("Alice")

def test_containing_no_match():
    assert not P.containing("li").test("Bob")

def test_containing_null_safe():
    assert not P.containing("li").test(None)

def test_not_containing():
    p = P.not_containing("li")
    assert p.test("Bob")
    assert not p.test("Alice")

def test_regexp_simple_match():
    assert P.regexp(r"[A-Z][a-z]+").test("Alice")

def test_regexp_no_match():
    assert not P.regexp(r"^\d+$").test("abc")

def test_regexp_partial_match():
    assert P.regexp(r"\d+").test("abc123def")

def test_regexp_null_safe():
    assert not P.regexp(r"\w+").test(None)

def test_negate_starting_with():
    p = P.starting_with("Al").negate()
    assert p.test("Bob")
    assert not p.test("Alice")

def test_predicate_in_filter():
    g = TinkerCat()
    try:
        for name in ["Alice", "Bob", "Alan", "Carol"]:
            g.add_vertex("person", name=name)
        p = P.starting_with("A")
        result = [v for v in g.vertices() if p.test(v.value("name"))]
        names = {v.value("name") for v in result}
        assert names == {"Alice", "Alan"}
    finally:
        g.close()

def test_ending_with_in_filter():
    g = TinkerCat()
    try:
        for email in ["alice@example.com", "bob@test.org", "carol@example.com"]:
            g.add_vertex("user", email=email)
        p = P.ending_with("@example.com")
        result = [v for v in g.vertices() if p.test(v.value("email"))]
        assert len(result) == 2
    finally:
        g.close()

def test_graphson_serialisation_starting_with():
    p = P.starting_with("Al")
    doc = {"@type": "g:P", "@value": {"predicate": "startingWith", "value": p.pattern}}
    text = json.dumps(doc)
    recovered = json.loads(text)
    assert recovered["@value"]["predicate"] == "startingWith"
    assert recovered["@value"]["value"] == "Al"

def test_graphson_serialisation_regexp():
    p = P.regexp(r"\d+")
    doc = {"@type": "g:P", "@value": {"predicate": "regexp", "value": p.pattern}}
    text = json.dumps(doc)
    recovered = json.loads(text)
    assert recovered["@value"]["predicate"] == "regexp"

File: tests/kotlin/test_7_7_2_p_predicates.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.collections.shouldHaveSize
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.7.2: P Predicate Additions.
 *
 * TinkerCat does not expose a built-in P class.  These tests implement the
 * same predicate semantics using Kotlin lambdas, demonstrating that the
 * underlying behaviour is correct and establishing contracts for a future
 * built-in P implementation.
 * Mirrors tests/python/test_7_7_2_p_predicates.py.
 */
class Test_7_7_2_PPredicates : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    // ---------------------------------------------------------------------------
    // Kotlin equivalents of P predicate factory methods
    // ---------------------------------------------------------------------------

    fun startingWith(prefix: String): (String?) -> Boolean = { it?.startsWith(prefix) ?: false }
    fun endingWith(suffix: String):  (String?) -> Boolean = { it?.endsWith(suffix)   ?: false }
    fun containing(sub: String):     (String?) -> Boolean = { it?.contains(sub)       ?: false }
    fun matchesRegex(pattern: String): (String?) -> Boolean {
        val re = Regex(pattern)
        return { it?.let { s -> re.containsMatchIn(s) } ?: false }
    }

    // ---------------------------------------------------------------------------
    // startingWith / notStartingWith
    // ---------------------------------------------------------------------------

    "startingWith matches correctly" {
        startingWith("Al")("Alice") shouldBe true
    }

    "startingWith does not match wrong prefix" {
        startingWith("Al")("Bob") shouldBe false
    }

    "startingWith is null-safe" {
        startingWith("Al")(null) shouldBe false
    }

    "not startingWith is logical negation" {
        val notAl: (String?) -> Boolean = { !startingWith("Al")(it) }
        notAl("Bob") shouldBe true
        notAl("Alice") shouldBe false
    }

    // ---------------------------------------------------------------------------
    // endingWith / notEndingWith
    // ---------------------------------------------------------------------------

    "endingWith matches correctly" {
        endingWith("son")("Jackson") shouldBe true
    }

    "endingWith does not match" {
        endingWith("son")("Alice") shouldBe false
    }

    "endingWith is null-safe" {
        endingWith("son")(null) shouldBe false
    }

    "not endingWith is logical negation" {
        val notSon: (String?) -> Boolean = { !endingWith("son")(it) }
        notSon("Alice") shouldBe true
        notSon("Jackson") shouldBe false
    }

    // ---------------------------------------------------------------------------
    // containing / notContaining
    // ---------------------------------------------------------------------------

    "containing matches substring" {
        containing("li")("Alice") shouldBe true
    }

    "containing does not match absent substring" {
        containing("li")("Bob") shouldBe false
    }

    "containing is null-safe" {
        containing("li")(null) shouldBe false
    }

    "not containing is logical negation" {
        val notLi: (String?) -> Boolean = { !containing("li")(it) }
        notLi("Bob") shouldBe true
        notLi("Alice") shouldBe false
    }

    // ---------------------------------------------------------------------------
    // regexp
    // ---------------------------------------------------------------------------

    "regexp matches capitalised name" {
        matchesRegex("[A-Z][a-z]+")("Alice") shouldBe true
    }

    "regexp does not match digits-only pattern against letters" {
        matchesRegex("^\\d+\$")("abc") shouldBe false
    }

    "regexp partial match works" {
        matchesRegex("\\d+")("abc123def") shouldBe true
    }

    "regexp is null-safe" {
        matchesRegex("\\w+")(null) shouldBe false
    }

    // ---------------------------------------------------------------------------
    // Predicates in graph filters
    // ---------------------------------------------------------------------------

    "startingWith used as graph filter" {
        for (name in listOf("Alice", "Bob", "Alan", "Carol")) {
            graph.addVertex(TinkerVertex.T.label, "person", "name", name)
        }
        val p = startingWith("A")
        val result = graph.vertices().asSequence()
            .filter { p(it.value<String>("name")) }.toList()
        result.map { it.value<String>("name") }.toSet() shouldBe setOf("Alice", "Alan")
    }

    "endingWith used as graph filter" {
        for (email in listOf("alice@example.com", "bob@test.org", "carol@example.com")) {
            graph.addVertex(TinkerVertex.T.label, "user", "email", email)
        }
        val p = endingWith("@example.com")
        val result = graph.vertices().asSequence()
            .filter { p(it.value<String>("email")) }.toList()
        result shouldHaveSize 2
    }

    "regexp used as graph filter" {
        for (name in listOf("alice", "Bob", "CAROL", "dave123")) {
            graph.addVertex(TinkerVertex.T.label, "user", "name", name)
        }
        val p = matchesRegex("^[A-Z]")
        val result = graph.vertices().asSequence()
            .filter { p(it.value<String>("name")) }.toList()
        result.map { it.value<String>("name") }.toSet() shouldBe setOf("Bob", "CAROL")
    }
})
